src/protocol/taiko_alethia/ProverManager.sol

```diff
@@ -0,0 +1,405 @@
+    function deposit() external payable {
+        balances[msg.sender] += msg.value;
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    /// @notice Withdraw available(unlocked) funds.
+    /// @param amount The amount to withdraw
+    function withdraw(uint256 amount) external {
+        balances[msg.sender] -= amount;
+
+        (bool ok,) = payable(msg.sender).call{value: amount}("");
+        require(ok, "Withdrawal failed");
+        emit Withdrawal(msg.sender, amount);
+    }
+
+    /// @inheritdoc IProposerFees
+    /// @dev This function advances to the next period if the current period has ended.
+    // TODO: deal with delayed publications
+    function payPublicationFee(address proposer, bool isDelayed) external payable {
+        require(msg.sender == inbox, "Only the Inbox contract can call this function");
+
+        // Accept additional deposit if sent
+        if (msg.value > 0) {
+            balances[proposer] += msg.value;
+            emit Deposit(proposer, msg.value);
+        }
+
+        uint256 _currentPeriod = currentPeriodId;
+
+        uint256 periodEnd = periods[_currentPeriod].end;
+        if (periodEnd != 0 && block.timestamp > periodEnd) {
+            // Advance to the next period
+            currentPeriodId++;
+            _currentPeriod++;
+            emit NewPeriod(_currentPeriod);
+        }
+
+        uint256 requiredFee = periods[_currentPeriod].fee;
+
+        // Deduct fee from proposer's balance and add to accumulated fees
+        balances[proposer] -= requiredFee;
+        periods[_currentPeriod].accumulatedFees += requiredFee;
+    }
+
+    /// @inheritdoc IProverManager
+    /// @dev The offered fee has to be at least `minUndercutPercentage` lower than the current best price.
+    /// @dev The current best price may be the current prover's fee or the fee of the next bid, depending on a few
+    /// conditions.
+    function bid(uint256 offeredFee) external {
+        uint256 _livenessBond = livenessBond;
+        require(balances[msg.sender] >= _livenessBond, "Insufficient balance for liveness bond");
+
+        uint256 currentPeriod = currentPeriodId;
+        Period storage _currentPeriod = periods[currentPeriod];
+        Period storage _nextPeriod = periods[currentPeriod + 1];
+        uint256 requiredMaxFee;
+        if (_currentPeriod.end == 0) {
+            // If the period is still active the bid has to be lower
+            uint256 currentFee = _currentPeriod.fee;
+            requiredMaxFee = currentFee - calculatePercentage(currentFee, minUndercutPercentage);
+            require(offeredFee <= requiredMaxFee, "Offered fee not low enough");
+
+            uint256 periodEnd = block.timestamp + succesionDelay;
+            _currentPeriod.end = periodEnd;
+            _currentPeriod.deadline = periodEnd + provingDeadline;
+        } else {
+            address _nextProverAddress = _nextPeriod.prover;
+            if (_nextProverAddress != address(0)) {
```

Comment:
Following the same logic of the my previous comment, there should be an abstraction for this, something like `_isBidded(address nextProver)` or similar. And then:

```suggestion
            if (_isBidded(_nextPeriod.prover)) {
```